File: sensor_monitor.js
// sensor_monitor.js — монитор датчиков (акселерометр, гироскоп, компас) на JavaScript (Node.js)

const readline = require('readline');

// ANSI-цвета
const colors = {
    cyan: '\x1b[96m',
    reset: '\x1b[0m',
    green: '\x1b[92m',
    yellow: '\x1b[93m',
    magenta: '\x1b[95m',
    blue: '\x1b[94m'
};

function clearScreen() {
    console.clear();
}

function updateData() {
    const t = Date.now() / 1000;
    return {
        accel: [
            0.2 * Math.sin(t * 0.5) + (Math.random() - 0.5) * 0.1,
            0.3 * Math.cos(t * 0.7) + (Math.random() - 0.5) * 0.1,
            9.81 + 0.1 * Math.sin(t * 0.3) + (Math.random() - 0.5) * 0.1
        ],
        gyro: [
            0.02 * Math.sin(t * 0.8) + (Math.random() - 0.5) * 0.02,
            0.03 * Math.cos(t * 0.6) + (Math.random() - 0.5) * 0.02,
            0.01 * Math.sin(t * 0.9) + (Math.random() - 0.5) * 0.02
        ],
        mag: [
            15.0 + 2.0 * Math.sin(t * 0.4) + (Math.random() - 0.5) * 1.0,
            -10.0 + 3.0 * Math.cos(t * 0.5) + (Math.random() - 0.5) * 1.0,
            40.0 + 4.0 * Math.sin(t * 0.6) + (Math.random() - 0.5) * 1.0
        ],
        angles: [
            2.0 * Math.sin(t * 0.2),
            1.5 * Math.cos(t * 0.25),
            45.0 + 10.0 * Math.sin(t * 0.1)
        ]
    };
}

function display(data) {
    clearScreen();
    console.log(`${colors.cyan}📡 SensorMonitor Pro — JavaScript Edition${colors.reset}`);
    console.log(`Акселерометр (м/с²):  X: ${data.accel[0].toFixed(2)}  Y: ${data.accel[1].toFixed(2)}  Z: ${data.accel[2].toFixed(2)}`);
    console.log(`Гироскоп (рад/с):     X: ${data.gyro[0].toFixed(2)}  Y: ${data.gyro[1].toFixed(2)}  Z: ${data.gyro[2].toFixed(2)}`);
    console.log(`Компас (мкТл):        X: ${data.mag[0].toFixed(2)}  Y: ${data.mag[1].toFixed(2)}  Z: ${data.mag[2].toFixed(2)}`);
    console.log(`Углы:  Roll: ${data.angles[0].toFixed(1)}°  Pitch: ${data.angles[1].toFixed(1)}°  Yaw: ${data.angles[2].toFixed(1)}°`);
    console.log(`\nНажмите Ctrl+C для выхода`);
}

function main() {
    const freq = 10;
    const interval = 1000 / freq;

    // Обработка Ctrl+C
    process.on('SIGINT', () => {
        console.log('\nВыход...');
        process.exit(0);
    });

    setInterval(() => {
        const data = updateData();
        display(data);
    }, interval);
}

main();

File: sensor_monitor.cpp
// sensor_monitor.cpp — монитор датчиков (акселерометр, гироскоп, компас) на C++

#include <iostream>
#include <cmath>
#include <thread>
#include <chrono>
#include <cstdlib>
#include <ctime>
#include <iomanip>
#include <vector>

using namespace std;

// ANSI-цвета для красивого вывода
#define RESET   "\033[0m"
#define RED     "\033[91m"
#define GREEN   "\033[92m"
#define YELLOW  "\033[93m"
#define BLUE    "\033[94m"
#define MAGENTA "\033[95m"
#define CYAN    "\033[96m"

struct SensorData {
    double accel[3];
    double gyro[3];
    double mag[3];
    double angles[3];
};

class SensorSimulator {
private:
    SensorData data;
    double freq;
    bool running;

public:
    SensorSimulator(double f = 10.0) : freq(f), running(true) {
        srand(time(nullptr));
    }

    void update() {
        double t = time(nullptr);
        // Акселерометр
        data.accel[0] = 0.2 * sin(t * 0.5) + (rand() % 100 - 50) / 1000.0;
        data.accel[1] = 0.3 * cos(t * 0.7) + (rand() % 100 - 50) / 1000.0;
        data.accel[2] = 9.81 + 0.1 * sin(t * 0.3) + (rand() % 100 - 50) / 1000.0;

        // Гироскоп
        data.gyro[0] = 0.02 * sin(t * 0.8) + (rand() % 100 - 50) / 5000.0;
        data.gyro[1] = 0.03 * cos(t * 0.6) + (rand() % 100 - 50) / 5000.0;
        data.gyro[2] = 0.01 * sin(t * 0.9) + (rand() % 100 - 50) / 5000.0;

        // Магнитометр
        data.mag[0] = 15.0 + 2.0 * sin(t * 0.4) + (rand() % 100 - 50) / 100.0;
        data.mag[1] = -10.0 + 3.0 * cos(t * 0.5) + (rand() % 100 - 50) / 100.0;
        data.mag[2] = 40.0 + 4.0 * sin(t * 0.6) + (rand() % 100 - 50) / 100.0;

        // Углы
        data.angles[0] = 2.0 * sin(t * 0.2);   // roll
        data.angles[1] = 1.5 * cos(t * 0.25);  // pitch
        data.angles[2] = 45.0 + 10.0 * sin(t * 0.1); // yaw
    }

    void run() {
        while (running) {
            update();
            display();
            this_thread::sleep_for(chrono::milliseconds((int)(1000.0 / freq)));
        }
    }

    void stop() { running = false; }

    void display() {
        // Очистка экрана (ANSI)
        cout << "\033[2J\033[H";
        cout << CYAN << "📡 SensorMonitor Pro — C++ Edition" << RESET << endl;
        cout << "Частота: " << freq << " Гц\n" << endl;
        cout << "Акселерометр (м/с²):  X: " << fixed << setprecision(2) << setw(6) << data.accel[0]
             << "  Y: " << setw(6) << data.accel[1] << "  Z: " << setw(6) << data.accel[2] << endl;
        cout << "Гироскоп (рад/с):     X: " << setw(6) << data.gyro[0]
             << "  Y: " << setw(6) << data.gyro[1] << "  Z: " << setw(6) << data.gyro[2] << endl;
        cout << "Компас (мкТл):        X: " << setw(6) << data.mag[0]
             << "  Y: " << setw(6) << data.mag[1] << "  Z: " << setw(6) << data.mag[2] << endl;
        cout << "Углы:  Roll: " << setw(6) << data.angles[0] << "°  Pitch: " << setw(6) << data.angles[1]
             << "°  Yaw: " << setw(6) << data.angles[2] << "°" << endl;
        cout << RESET << "\nНажмите Ctrl+C для выхода" << endl;
    }

    SensorData getData() const { return data; }
};

int main(int argc, char* argv[]) {
    double freq = 10.0;
    if (argc > 1) freq = atof(argv[1]);
    if (freq < 1) freq = 1;
    if (freq > 100) freq = 100;

    SensorSimulator sim(freq);
    // Запускаем в отдельном потоке (для возможности остановки по Ctrl+C)
    thread t(&SensorSimulator::run, &sim);

    // Ждём сигнала Ctrl+C
    cin.get();
    sim.stop();
    t.join();
    return 0;
}
